### PR TITLE
gh-202: Request List redesign

### PR DESCRIPTION
**Pull request by NigelPalmer**
gh-202 has redesigned the request list so that the data elements are now grouped under data classes, which in turn are grouped under data schemas. The following functionality needs to be checked:

1. Checking/unchecking "All selected" will select/deselect all the items.
2. Checking/unchecking a Schema will select/deselect all the items in that schema.
3. Checking/unchecking a Class will select/deselect all the items in that class.
4. Clicking "remove selected" does delete the selected items.
5. All delete buttons delete the item that they are in line with.
6. If you delete a schema then all classes and elements belonging to that schema are deleted.
7. If you delete a class then all elements belonging to that class are deleted.
8. If a class only has one element then when that element is deleted, the class is deleted as well.
9. If a schema has only one class, when deleted the class is deleted, the schema is deleted as well.
10. Data elements are deleted from the queries when the elements are deleted.
11. All copying of elements to other requests works correctly. Schemas, classes and elements can all be copied around. 
12. Schemas and classes can be collapsed/expanded in the UI.

Follow up Issues will be created to address the following known issues:
1. The green tick that appears to say that an element has been referenced by a request is no longer refreshing correctly.
2. The list of elements on the template screen is not grouped by Schema and Class.
3. Tests of the affected components need to be implemented.

**Additional Notes:**
Before this work was done the list of Data Elements in a Request was shown to the user by mapping the elements in the request to the Root model, and showing those Root model elements. I believe this was done to allow for data elements to be added and removed from the Request, using the copy menu, without any errors occurring. But this approach added an extra layer of complexity for deleting Schemas, Classes and Elements. Therefore, I changed the code so the Request itself is loaded and then manipulated directly. The issues with added and removing elements from the Request itself was dealt with by falling back to the Root model for those actions only. This allows for simpler code that will be easier to maintain moving forwards.

### CODE REVIEW

File: src/app/data-explorer/data-schema-row/data-schema-row.component.html [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      ><!--(createRequestClicked)="handleCreateRequest($event)"-->
+      <button
+        mat-button
+        class="mdm-data-schema-row__header-button-updown fa-solid"
+        [ngClass]="{ 'fa-angle-up': visible, 'fa-angle-down': !visible }"
+        (click)="toggleCollapse()"
+      ></button>
+    </span>
+  </div>
+  <div [hidden]="!visible" class="mdm-my-request__request-elements" cCollapse>
```
Comment (pjmonks) [inline src/app/data-explorer/data-schema-row/data-schema-row.component.html:58]: What is the `cCollapse` attribute for? A typo?

File: src/app/data-explorer/data-schema-row/data-schema-row.component.ts [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+} from '@angular/core';
+import { MatCheckboxChange } from '@angular/material/checkbox';
+import { RequestElementAddDeleteEvent } from 'src/app/shared/data-element-in-request/data-element-in-request.component';
+import {
+  DataClassWithElements,
+  DataElementSearchResult,
+  DataItemDeleteEvent,
+  DataRequestQueryType,
+  DataSchema,
+  // RefreshRequestEvent,
```
Comment (pjmonks) [inline src/app/data-explorer/data-schema-row/data-schema-row.component.ts:36]: Remove commented out code

File: src/app/data-explorer/data-schema-row/data-schema-row.component.ts [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+  @Input() requestName = '';
+  @Input() requestId = '';
+  @Input() suppressViewRequestsDialogButton = false;
+  @Input() canDelete = true;
+  @Input() sourceTargetIntersections: DataAccessRequestsSourceTargetIntersections;
+  @Input() allSelected?: SelectionChange;
+
+  @Output() deleteItemEvent = new EventEmitter<DataItemDeleteEvent>();
+  @Output() requestAddDelete = new EventEmitter<RequestElementAddDeleteEvent>();
+  // @Output() requestCreated = new EventEmitter<CreateRequestEvent>();
```
Comment (pjmonks) [inline src/app/data-explorer/data-schema-row/data-schema-row.component.ts:59]: Remove commented out code

File: src/app/data-explorer/data-schema-row/data-schema-row.component.ts [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+  dataQueryType: DataRequestQueryType = 'data';
+
+  constructor(private dataSchemaService: DataSchemaService) {
+    this.sourceTargetIntersections = {
+      dataAccessRequests: [],
+      sourceTargetIntersections: [],
+    };
+  }
+
+  // Init, get the classes from the request. and list.
```
Comment (pjmonks) [inline src/app/data-explorer/data-schema-row/data-schema-row.component.ts:81]: I feel this comment is pointing out the obvious a bit. Another way to make the function clearer is to rename the service method to `this.dataSchemaService.getDataSchemaElements()`

File: src/app/data-explorer/data-schema-row/data-schema-row.component.ts [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+      return;
+    }
+
+    this.selectSchema(event.checked, {
+      instigator: 'parent',
+    });
+
+    this.schemaCheckedEvent.emit();
+    /*
+    this.schemaCheckedEvent.emit({
```
Comment (pjmonks) [inline src/app/data-explorer/data-schema-row/data-schema-row.component.ts:132]: Remove commented out code

File: src/app/mauro/data-schema-service.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+} from '../data-explorer/data-explorer.types';
+import { DataModelService } from './data-model.service';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class DataSchemaService {
+  constructor(private dataModels: DataModelService) {}
+
+  dataRequestClasses(dataSchemas: DataSchema[]): DataClassWithElements[] {
```
Comment (pjmonks) [inline src/app/mauro/data-schema-service.ts:36]: I think the 3 functions `dataRequestClasses()`, `dataRequestElements()` and `dataSchemaElements()` could have better names, I'm not sure what they are doing on first glance.

File: src/app/mauro/sort.service.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+@Injectable({
+  providedIn: 'root',
+})
+export class SortService {
+  ascString(a: string, b: string): number {
```
Comment (pjmonks) [inline src/app/mauro/sort.service.ts:25]: These two functions look a bit too simple to include in an injectable service. Have you considered just making them pure TypeScript functions?

File: src/app/pages/data-element/data-element.component.ts [hunk 8]
```diff
@@ -100,7 +100,8 @@ export class DataElementComponent implements OnInit {
         ([dataElementDetail, isBookmarked, profile, sourceTargetIntersections]) => {
           this.dataElement = dataElementDetail;
           this.isBookmarked = isBookmarked;
-          this.dataElementSearchResult = {
+          this.dataElementSearchResult = [];
+          this.dataElementSearchResult.push({
```
Comment (pjmonks) [inline src/app/pages/data-element/data-element.component.ts:104]: Why not just `this.dataElementSearchResult = [{ ... }]`?

File: src/app/pages/my-request-detail/my-request-detail.component.ts [hunk 9]
```diff
@@ -207,21 +238,135 @@ export class MyRequestDetailComponent implements OnInit {
       });
   }
 
-  removeItem(event: DataElementDeleteEvent) {
-    const item = event.item;
-    this.removeOneOrMoreDataElements([item]);
+  removeItem(event: DataItemDeleteEvent) {
+    if (!event.dataSchema) {
+      this.toastr.error('Data schema undefined', 'Unable to delete items');
+    } else {
```
Comment (pjmonks) [inline src/app/pages/my-request-detail/my-request-detail.component.ts:244]: I would suggest less indentation here, my personal code style tends to try and avoid if-elses and return as soon as possible, for instance:

```ts
if (!event.dataSchema) {
  this.toastr.error();
  return;
}

if (!userFacingText) {
  this.toastr.error();
  return;
}

// ... and so on
```

File: src/app/pages/my-request-detail/my-request-detail.component.html [hunk 10]
```diff
@@ -55,41 +55,46 @@ <h2>Queries</h2>
...
+        >Select all</mat-checkbox
+      >
+      <button
+        class="ms-4"
+        mat-raised-button
+        color="primary"
+        [disabled]="removeSelectedButtonDisabled"
+        (click)="removeSelected()"
+      >
+        Remove selected ...
```
Comment (pjmonks) [inline src/app/pages/my-request-detail/my-request-detail.component.html:75]: Might be a bug with the "Remove selected..." button, looking at a "submitted" request, I can get in this state:

image

The individual remove buttons on each row are hidden at least, but the selection area remove button isn't.

File: src/app/pages/my-request-detail/my-request-detail.component.ts [hunk 11]
```diff
@@ -234,201 +379,87 @@ export class MyRequestDetailComponent implements OnInit {
...
-    queryType: DataRequestQueryType
-  ): Observable<DataRequestQueryPayload> {
-    if (!this.request?.id) {
-      return EMPTY;
-    }
+  /**
+   * Methods for general page management
+   */
+  private setBackButtonProperties() {
+    this.backRouterLink = '/requests';
```
Comment (pjmonks) [inline src/app/pages/my-request-detail/my-request-detail.component.ts:407]: I have a feeling I removed `backRouterLink` and `backLabel` from a previous PR. They didn't seem necessary because the back link on the page always is the same. Double check this wasn't re-introduced in a rebase.

File: src/app/shared/data-element-in-request/data-element-in-request.component.spec.ts [hunk 12]
```diff
@@ -107,7 +101,7 @@ describe('DataElementInRequestComponent', () => {
   it('should create', () => {
     expect(harness?.isComponentCreated).toBeTruthy();
   });
-
+  /*
```
Comment (pjmonks) [inline src/app/shared/data-element-in-request/data-element-in-request.component.spec.ts:104]: Revisit these tests, or remove them if commented out.

File: src/app/shared/data-element-in-request/data-element-in-request.component.ts [hunk 13]
```diff
@@ -110,6 +116,12 @@ export class DataElementInRequestComponent implements OnInit, OnDestroy {
       return;
     }
 
+    /*
+    if (this.dataElement && !this.dataElements) {
```
Comment (pjmonks) [inline src/app/shared/data-element-in-request/data-element-in-request.component.ts:120]: Remove commented out code

File: src/app/shared/data-element-in-request/data-element-in-request.component.ts [hunk 14]
```diff
@@ -128,56 +140,124 @@ export class DataElementInRequestComponent implements OnInit, OnDestroy {
...
+      dataElements = this.dataElements;
+
+      this.broadcast.loading({ isLoading: true, caption: 'Updating your request...' });
+
+      // All the elements will belong to the same dataModel so we can just check the first one.
+      // const dataElement: DataElementSearchResult = this.dataElements[0];
+
+      // If the dataModel and target is the same set the dataModel to be the root
+      /*
+      if (dataElements[0].model === targetDataModelId) {
```
Comment (pjmonks) [inline src/app/shared/data-element-in-request/data-element-in-request.component.ts:160]: Remove commented out code

File: src/app/shared/data-element-in-request/data-element-in-request.component.ts [hunk 15]
```diff
@@ -267,30 +371,38 @@ export class DataElementInRequestComponent implements OnInit, OnDestroy {
...
-      dataElementIds.push(this.dataElement.id);
-    }
+  private dataElementSearchResultsToDataElementDTOs(
+    dataElements: DataElementSearchResult[]
+  ): DataElementDto[] {
+    return dataElements.map((element) => {
+      return (
+        element
+          ? {
+              // ...element
```
Comment (pjmonks) [inline src/app/shared/data-element-in-request/data-element-in-request.component.ts:381]: Remove commented out code

**Review by pjmonks — CHANGES_REQUESTED** [review]
I think I've checked everything now - it's a big PR! As far as I can tell I cannot break most of the functionality as listed in the PR description, so 👍 Only spotted one functional bug to do with the "Remove selected" button, which is listed in the feedback comments.

I can see a lot of effort went into this and can definitely see some improvements. Nearly all the feedback I've left is related to code cleanup after such a large change, if this could be reviewed please.

File: src/app/shared/data-element-in-request/data-element-in-request.component.html [hunk 16]
```diff
@@ -17,9 +17,14 @@
...
-  <button #matMenuTrigger mat-raised-button color="primary" [matMenuTriggerFor]="menu">
-    {{ caption }}
-  </button>
+  <button
+    #matMenuTrigger
+    mat-raised-button
+    color="primary"
+    [matMenuTriggerFor]="menu"
+    class="fa-solid fa-add"
+    title="{{ caption }}"
```
[the remark this answers is not in the document]
Comment (NigelPalmer) [reply src/app/shared/data-element-in-request/data-element-in-request.component.html:26]: Updated as suggested.

File: src/app/shared/data-element-in-request/data-element-in-request.component.html [hunk 17]
```diff
@@ -17,9 +17,14 @@
 SPDX-License-Identifier: Apache-2.0
 -->
 <ng-container>
-  <button #matMenuTrigger mat-raised-button color="primary" [matMenuTriggerFor]="menu">
-    {{ caption }}
-  </button>
+  <button
+    #matMenuTrigger
```
[the remark this answers is not in the document]
Comment (NigelPalmer) [reply src/app/shared/data-element-in-request/data-element-in-request.component.html:21]: aria-label added.

File: src/app/pages/my-request-detail/my-request-detail.component.html [hunk 18]
```diff
@@ -55,41 +55,46 @@ <h2>Queries</h2>
...
 
     <h2>Data elements</h2>
-
     <p *ngIf="request.status === 'unsent'">
       You can add more data elements to this request by
       <a routerLink="/browse">browsing</a> or <a routerLink="/search">searching</a> our
       catalogue.
     </p>
 
+    <div class="mdm-my-request__request-elements-footer">
```
[the remark this answers is not in the document]
Comment (NigelPalmer) [reply src/app/pages/my-request-detail/my-request-detail.component.html:64]: I have renamed the CSS class to &__request-elements-selectall since it is only applied to the selectall region of the page. I have also updated the css to include an indentation so that the "select all" checkbox lines up with everything below it.

File: src/app/data-explorer/data-class-row/data-class-row.component.ts [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+  Output,
+  SimpleChanges,
+} from '@angular/core';
+import { MatCheckboxChange } from '@angular/material/checkbox';
+import { RequestElementAddDeleteEvent } from 'src/app/shared/data-element-in-request/data-element-in-request.component';
+import {
+  DataClassWithElements,
+  DataElementSearchResult,
+  DataItemDeleteEvent,
+  // RefreshRequestEvent,
```
[the remark this answers is not in the document]
Comment (NigelPalmer) [reply src/app/data-explorer/data-class-row/data-class-row.component.ts:34]: Removed

File: src/app/data-explorer/data-element-row/data-element-row.component.html [hunk 20]
```diff
@@ -38,27 +46,24 @@ <h3>{{ item.label }}</h3>
       <button
         *ngIf="canDelete"
         mat-raised-button
-        class="mdm-data-element-row__header-button"
+        class="mdm-data-element-row__header-button fa fa-trash-alt"
         color="primary"
         (click)="removeElement()"
-      >
-        Remove
```
[the remark this answers is not in the document]
Comment (NigelPalmer) [reply src/app/data-explorer/data-element-row/data-element-row.component.html:45]: Updated as suggested.

File: src/app/data-explorer/data-element-row/data-element-row.component.ts [hunk 21]
```diff
@@ -16,26 +16,64 @@ limitations under the License.
...
+  @Input() classSelected?: SelectionChange;
 
-  @Output() delete = new EventEmitter<DataElementDeleteEvent>();
+  @Output() deleteItemEvent = new EventEmitter<DataItemDeleteEvent>();
   @Output() requestAddDelete = new EventEmitter<RequestElementAddDeleteEvent>();
   @Output() requestCreated = new EventEmitter<CreateRequestEvent>();
+  @Output() setRemoveSelectedButtonDisabledEvent = new EventEmitter();
+
+  padding = 'default';
+  itemAsArray: DataElementSearchResult[] = [] as DataElementSearchResult[];
```
[the remark this answers is not in the document]
Comment (NigelPalmer) [reply src/app/data-explorer/data-element-row/data-element-row.component.ts:59]: Casting removed.

File: src/app/data-explorer/data-element-row/data-element-row.component.ts [hunk 22]
```diff
@@ -16,26 +16,64 @@ limitations under the License.
...
+
+  override ngOnInit(): void {
+    this.padding = this.nestedPadding ? 'nested' : 'default';
+    if (this.item) {
+      this.itemAsArray.push(this.item);
```
[the remark this answers is not in the document]
Comment (NigelPalmer) [reply src/app/data-explorer/data-element-row/data-element-row.component.ts:64]: An array is required to be passed into the `mdm-data-element-in-request` component, since this now accepts multiple `dataElements` from other controls. But, the simple syntax of putting the item in square brackets means that this can simply be done in the html when the item is assigned to the `dataElements `parameter. Therefore, the `itemAsArray `code has been deleted.

File: src/app/data-explorer/data-element-search-result/data-element-search-result.component.ts [hunk 23]
```diff
@@ -59,6 +61,12 @@ export class DataElementSearchResultComponent {
     };
   }
 
+  ngOnInit(): void {
+    if (this.item) {
+      this.dataElementSearchResults.push(this.item);
```
[the remark this answers is not in the document]
Comment (NigelPalmer) [reply src/app/data-explorer/data-element-search-result/data-element-search-result.component.ts:66]: Again, code not required. The html can simply reference [item]. Code has been deleted.

File: src/app/data-explorer/data-requests.service.ts [hunk 24]
```diff
@@ -264,6 +266,110 @@ export class DataRequestsService {
...
+   * @param items: The data elements to be removed
+   *
+   * @returns: Observable of DataElementMultipleOperationResult
+   */
+  deleteDataClassMultiple(
+    dataClasses: DataClass[]
+  ): Observable<DataElementMultipleOperationResult> {
+    return of(dataClasses).pipe(
+      switchMap((dataClassArray: DataClass[]) => from(dataClassArray)),
+      filter((dataClass) => dataClass !== null),
```
[the remark this answers is not in the document]
Comment (NigelPalmer) [reply src/app/data-explorer/data-requests.service.ts:282]: That is useful to know. Updated as suggested.

File: src/app/data-explorer/data-schema-row/data-schema-row.component.html [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+      <mdm-data-element-in-request
+        class="mdm-data-schema-row__header-button"
+        *ngIf="sourceTargetIntersections.dataAccessRequests.length > 0"
+        [dataElements]="schemaElements"
+        [sourceTargetIntersections]="sourceTargetIntersections"
+        [caption]="'Add to other requests'"
+        [suppressViewRequestsDialogButton]="suppressViewRequestsDialogButton"
+        (requestAddDelete)="handleRequestAddDelete($event)"
+      ></mdm-data-element-in-request
+      ><!--(createRequestClicked)="handleCreateRequest($event)"-->
```
[the remark this answers is not in the document]
Comment (NigelPalmer) [reply src/app/data-explorer/data-schema-row/data-schema-row.component.html:49]: Removed.

File: src/app/data-explorer/data-schema-row/data-schema-row.component.html (hunk 1 above)
Comment (NigelPalmer) [reply src/app/data-explorer/data-schema-row/data-schema-row.component.html:58]: This is a hangover from my initial attempts to expand and collapse sections. Deleted.

File: src/app/data-explorer/data-schema-row/data-schema-row.component.ts (hunk 2 above)
Comment (NigelPalmer) [reply src/app/data-explorer/data-schema-row/data-schema-row.component.ts:36]: Removed

File: src/app/data-explorer/data-schema-row/data-schema-row.component.ts (hunk 3 above)
Comment (NigelPalmer) [reply src/app/data-explorer/data-schema-row/data-schema-row.component.ts:59]: Removed

File: src/app/data-explorer/data-schema-row/data-schema-row.component.ts (hunk 4 above)
Comment (NigelPalmer) [reply src/app/data-explorer/data-schema-row/data-schema-row.component.ts:81]: The comment is wrong. The line of code that follows it doesn't even do what the comments says. Comment deleted and service methods have been renamed to be prefixed with "get" to make things clearer as suggested.

File: src/app/data-explorer/data-schema-row/data-schema-row.component.ts (hunk 5 above)
Comment (NigelPalmer) [reply src/app/data-explorer/data-schema-row/data-schema-row.component.ts:132]: Removed.

File: src/app/mauro/data-schema-service.ts (hunk 6 above)
Comment (NigelPalmer) [reply src/app/mauro/data-schema-service.ts:36]: I renamed these to include the prefix "get" as suggested in a couple of comments above.

File: src/app/shared/data-element-in-request/data-element-in-request.component.ts (hunk 15 above)
Comment (NigelPalmer) [reply src/app/shared/data-element-in-request/data-element-in-request.component.ts:381]: Removed

File: src/app/shared/data-element-in-request/data-element-in-request.component.ts (hunk 14 above)
Comment (NigelPalmer) [reply src/app/shared/data-element-in-request/data-element-in-request.component.ts:160]: Removed

File: src/app/shared/data-element-in-request/data-element-in-request.component.ts (hunk 13 above)
Comment (NigelPalmer) [reply src/app/shared/data-element-in-request/data-element-in-request.component.ts:120]: Removed

File: src/app/shared/data-element-in-request/data-element-in-request.component.spec.ts (hunk 12 above)
Comment (NigelPalmer) [reply src/app/shared/data-element-in-request/data-element-in-request.component.spec.ts:104]: These commented out tests will be restored and modified by gh-244. I have added a comment to clarify why they are still in  the code base. I don't want to delete them since I think it will make the test writers job easier if they are still here.

File: src/app/pages/my-request-detail/my-request-detail.component.ts (hunk 11 above)
Comment (NigelPalmer) [reply src/app/pages/my-request-detail/my-request-detail.component.ts:407]: They got half restored in conflict management and my investigations into whether they were required or not was inconclusive. So I put them back to be on the safer side. Now removed.

File: src/app/pages/my-request-detail/my-request-detail.component.html (hunk 10 above)
Comment (NigelPalmer) [reply src/app/pages/my-request-detail/my-request-detail.component.html:75]: It doesn't make sense for the button to be visible, or the checkboxes to be visible when a request has been submitted. I have updated the code to remove the check boxes. Since I had done that, I also went ahead and updated the Templates page so that it uses the new data list of elements, which removes the need for a follow-up task.

File: src/app/pages/my-request-detail/my-request-detail.component.ts (hunk 9 above)
Comment (NigelPalmer) [reply src/app/pages/my-request-detail/my-request-detail.component.ts:244]: Updated as suggested

File: src/app/mauro/sort.service.ts (hunk 7 above)
Comment (NigelPalmer) [reply src/app/mauro/sort.service.ts:25]: I've changed to a static class, but I'm not 100% sure that is what was intended with this change.

File: src/app/pages/data-element/data-element.component.ts (hunk 8 above)
Comment (NigelPalmer) [reply src/app/pages/data-element/data-element.component.ts:104]: Because in my rush I forgot arrays could be initialised like that. Fixed now.

File: src/app/data-explorer/data-class-row/data-class-row.component.ts [hunk 26]
```diff
@@ -0,0 +1,172 @@
...
+  removeClass() {
+    if (this.dataClassWithElements) {
+      this.deleteItemEvent.emit({
+        dataClassWithElements: this.dataClassWithElements,
+      });
+    }
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  classChecked(event: MatCheckboxChange) {
```
[the remark this answers is not in the document]
Comment (NigelPalmer) [reply src/app/data-explorer/data-class-row/data-class-row.component.ts:120]: Updated as suggested.

File: src/app/mauro/sort.type.ts [hunk 27]
```diff
@@ -18,15 +16,12 @@ limitations under the License.
...
-})
-export class SortService {
-  ascString(a: string, b: string): number {
+export class Sort {
+  static ascString(a: string, b: string): number {
     return a > b ? 1 : b > a ? -1 : 0;
   }
 
-  descString(a: string, b: string): number {
+  static descString(a: string, b: string): number {
```
Comment (pjmonks) [inline src/app/mauro/sort.type.ts:24]: When I suggested this change, I actually meant making plain functions because I felt like a class was unecessary, like this:

```ts
export const compareAsc = (a: string, b: string) => {
  return a > b ? 1 : b > a ? -1 : 0;
}
```

But this is fine for now.

File: src/setupJest.ts [hunk 28]
```diff
@@ -16,5 +16,5 @@ limitations under the License.
 
 SPDX-License-Identifier: Apache-2.0
 */
-// import 'jest-preset-angular/setup-jest';
+import 'jest-preset-angular/setup-jest'; // This allows jest tests to run in Visual Studio Code
```
Comment (pjmonks) [inline src/setupJest.ts:19]: This I'm confused by - but not enough to block the PR...

I remember having to comment this out once because I think the `jest-preset-angular` setup ended up being called twice for some reason, related to the Angular dev builders in `angular.json` file. Tests still run though now, so why was this added back in? (For curiosity only)

**Review by pjmonks — APPROVED** [review]
All feedback addressed, everything looks good to go, well done 🎉 

Approved ✔️

File: src/app/mauro/sort.type.ts (hunk 27 above)
Comment (NigelPalmer) [reply src/app/mauro/sort.type.ts:24]: If we add other sort methods, then this list will grow, so I think it makes sense to keep them together in a class.

File: src/setupJest.ts (hunk 28 above)
Comment (NigelPalmer) [reply src/setupJest.ts:19]: I installed "Jest Test Explorer" into Visual Studio Code to enable me to easily step through a test to debug it. The test explorer was unable to run the jest tests until I restored that line of code. Obviously my comment was not as clear as it should have been.
